Hibernate Subselect Entity not supported by EntityMetamodelImpl

<!--- This template is for bugs. Remove it for other issues. -->

<!--- Choose an expressive title -->
### Description
After playing a bit with table functions (see also #181), I had the idea to experiment with mapping a `generate_series` query to an entity using the [`@Subselect`](https://docs.jboss.org/hibernate/orm/5.2/javadocs/org/hibernate/annotations/Subselect.html) annotation. Find below the produced stack trace.

A trivial work around is to wrap the query into to a view and map that instead. I am uncertain whether a fix for this is worthwhile, but it could perhaps be a step towards implementing table functions (#181).



<!--- Explain what you did and maybe show some code excerpts -->
```java
package com.pallasathenagroup.entities.series;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.hibernate.annotations.Immutable;
import org.hibernate.annotations.Subselect;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import java.time.LocalDateTime;

@Data
@Entity
@Immutable
@Subselect("select start_date_time, start_date_time + interval '1 year' as end_date_time " +
    "from generate_series('2016-09-01'::timestamp, date_trunc('year', now()) + interval '8 months', '1 year') as start_date_time")
@EqualsAndHashCode(of = {"start"})
public class CollegeYearSeries implements Comparable<CollegeYearSeries> {

    @Id
    @Column(name = "start_date_time")
    private LocalDateTime start;

    @Column(name = "end_date_time")
    private LocalDateTime end;

    @Override
    public int compareTo(CollegeYearSeries o) {
        return getStart().compareTo(o.getStart());
    }
}
```


```
Caused by: java.lang.NullPointerException
	at com.blazebit.persistence.impl.hibernate.HibernateExtendedQuerySupport.getColumnTypes(HibernateExtendedQuerySupport.java:246)
	at com.blazebit.persistence.impl.EntityMetamodelImpl.<init>(EntityMetamodelImpl.java:110)
	at com.blazebit.persistence.impl.CriteriaBuilderFactoryImpl.<init>(CriteriaBuilderFactoryImpl.java:81)
	at com.blazebit.persistence.impl.CriteriaBuilderConfigurationImpl.createCriteriaBuilderFactory(CriteriaBuilderConfigurationImpl.java:599)
	at com.pallasathenagroup.iris.producers.CriteriaBuilderFactoryProducer.init(CriteriaBuilderFactoryProducer.java:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.as.ee.component.ManagedReferenceLifecycleMethodInterceptor.processInvocation(ManagedReferenceLifecycleMethodInterceptor.java:96)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.ejb.Jsr299BindingsInterceptor.doLifecycleInterception(Jsr299BindingsInterceptor.java:114)
	at org.jboss.as.weld.ejb.Jsr299BindingsInterceptor.processInvocation(Jsr299BindingsInterceptor.java:103)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.invocation.InterceptorContext$Invocation.proceed(InterceptorContext.java:437)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
	at org.jboss.as.weld.ejb.EjbRequestScopeActivationInterceptor.processInvocation(EjbRequestScopeActivationInterceptor.java:83)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.injection.WeldInjectionInterceptor.processInvocation(WeldInjectionInterceptor.java:53)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.ManagedReferenceFieldInjectionInterceptorFactory$ManagedReferenceFieldInjectionInterceptor.processInvocation(ManagedReferenceFieldInjectionInterceptorFactory.java:107)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.AroundConstructInterceptorFactory$1.processInvocation(AroundConstructInterceptorFactory.java:28)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.injection.WeldInterceptorInjectionInterceptor.processInvocation(WeldInterceptorInjectionInterceptor.java:56)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.ejb.Jsr299BindingsCreateInterceptor.processInvocation(Jsr299BindingsCreateInterceptor.java:100)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.NamespaceContextInterceptor.processInvocation(NamespaceContextInterceptor.java:50)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInOurTx(CMTTxInterceptor.java:275)
	... 29 more
```
### Expected behavior
<!--- What outcome would you expect? -->

### Actual behavior
<!--- What happened that you didn't expect? -->

### Steps to reproduce
<!--- Give us enough details so we can create a testcase that reproduces this problem -->
<!--- Idealy you would create a test case based on the template project from: https://github.com/Blazebit/blaze-persistence-test-case-template -->
<!--- Either attach a ZIP containing the test case or create a pull request against the test-case-template repository and put the link to the PR here -->

### Environment
<!--- Environment info like e.g.					-->
<!--- Version:            1.2.0-Alpha1				-->
<!--- JPA-Provider:       Hibernate 5.2.7.Final		-->
<!--- DBMS:               PostgreSQL 9.6.2			-->
<!--- Application Server: Java SE					-->

Version:            1.2.0.Alpha3
JPA-Provider:       Hibernate 5.2.12
DBMS:               PostgresSQL
Application Server: Wildfly

--- a/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java
+++ b/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java
@@ -1,0 +1,117 @@
+/*
+ * Copyright 2014 - 2018 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.testsuite;
+
+import com.blazebit.persistence.CriteriaBuilder;
+import org.hibernate.annotations.Generated;
+import org.hibernate.annotations.GenerationTime;
+import org.hibernate.annotations.Subselect;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.persistence.Entity;
+import javax.persistence.Table;
+import javax.persistence.Column;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.OneToOne;
+import javax.persistence.PrimaryKeyJoinColumn;
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+
+public class Issue519Test extends AbstractCoreTest {
+
+    A instance;
+
+    @Before
+    public void setUp() throws Exception {
+        instance = new A();
+        em.persist(instance);
+        em.refresh(instance);
+    }
+
+    @Test
+    public void subselectEntityTest() {
+        CriteriaBuilder<A> a1 = cbf.create(em, A.class)
+                .from(A.class, "a")
+                .where("a.b.b").eq(5l);
+
+        String queryString = a1.getQueryString();
+        List<A> resultList = a1.getResultList();
+        assertTrue(resultList.contains(instance));
+    }
+
+    @Entity
+    @Table(name = "A")
+    public static class A {
+
+        @Id
+        @GeneratedValue
+        private Long id;
+
+        @Generated(GenerationTime.INSERT)
+        @OneToOne(mappedBy = "a")
+        private B b;
+
+
+    }
+
+    @Entity
+    @Subselect("SELECT 5 as b, a.id AS aId FROM A")
+    public static class B {
+
+        private Long aId;
+
+        private A a;
+
+        private Long b;
+
+        @Id
+        public Long getaId() {
+            return aId;
+        }
+
+        public void setaId(Long aId) {
+            this.aId = aId;
+        }
+
+        @OneToOne
+        @PrimaryKeyJoinColumn(name = "a_id")
+        public A getA() {
+            return a;
+        }
+
+        public void setA(A a) {
+            this.a = a;
+        }
+
+        @Column
+        public Long getB() {
+            return b;
+        }
+
+        public void setB(Long b) {
+            this.b = b;
+        }
+    }
+
+    @Override
+    protected Class<?>[] getEntityClasses() {
+        return new Class<?>[] { A.class, B.class };
+    }
+}

--- a/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java
+++ b/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java
@@ -17,43 +17,60 @@
 package com.blazebit.persistence.testsuite;
 
 import com.blazebit.persistence.CriteriaBuilder;
+import com.blazebit.persistence.testsuite.tx.TxVoidWork;
 import org.hibernate.annotations.Generated;
 import org.hibernate.annotations.GenerationTime;
 import org.hibernate.annotations.Subselect;
-import org.junit.Before;
 import org.junit.Test;
 
-import javax.persistence.Entity;
-import javax.persistence.Table;
 import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.EntityManager;
 import javax.persistence.GeneratedValue;
 import javax.persistence.Id;
 import javax.persistence.OneToOne;
 import javax.persistence.PrimaryKeyJoinColumn;
+import javax.persistence.Table;
 import java.util.List;
 
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
+/**
+ * @author Jan-Willem Gmelig Meyling
+ * @since 1.2.0
+ */
 public class Issue519Test extends AbstractCoreTest {
 
-    A instance;
+    private A instance;
 
-    @Before
-    public void setUp() throws Exception {
-        instance = new A();
-        em.persist(instance);
-        em.refresh(instance);
+    @Override
+    protected Class<?>[] getEntityClasses() {
+        return new Class<?>[]{ A.class, B.class };
+    }
+
+    @Override
+    protected void setUpOnce() {
+        transactional(new TxVoidWork() {
+            @Override
+            public void work(EntityManager em) {
+                instance = new A();
+                em.persist(instance);
+                em.flush();
+                em.refresh(instance);
+            }
+        });
     }
 
     @Test
     public void subselectEntityTest() {
-        CriteriaBuilder<A> a1 = cbf.create(em, A.class)
+        CriteriaBuilder<A> cb = cbf.create(em, A.class)
                 .from(A.class, "a")
-                .where("a.b.b").eq(5l);
+                .where("a.b.b").eq(5L);
 
-        String queryString = a1.getQueryString();
-        List<A> resultList = a1.getResultList();
-        assertTrue(resultList.contains(instance));
+        List<A> resultList = cb.getResultList();
+        assertEquals(1, resultList.size());
+        assertEquals(Long.valueOf(5L), resultList.get(0).b.b);
     }
 
     @Entity
@@ -68,11 +85,10 @@ public class Issue519Test extends AbstractCoreTest {
         @OneToOne(mappedBy = "a")
         private B b;
 
-
     }
 
     @Entity
-    @Subselect("SELECT 5 as b, a.id AS aId FROM A")
+    @Subselect("SELECT 5 as b, a.id AS aId FROM A a")
     public static class B {
 
         private Long aId;
@@ -108,10 +124,5 @@ public class Issue519Test extends AbstractCoreTest {
         public void setB(Long b) {
             this.b = b;
         }
-    }
-
-    @Override
-    protected Class<?>[] getEntityClasses() {
-        return new Class<?>[] { A.class, B.class };
     }
 }

--- a/core/testsuite/src/test/java/com/blazebit/persistence/testsuite/Issue521Test.java
+++ b/core/testsuite/src/test/java/com/blazebit/persistence/testsuite/Issue521Test.java
@@ -28,12 +28,12 @@ public class Issue521Test extends AbstractCoreTest {
     @Test
     public void secondaryTableWithJoinedInheritanceTest() {
         SecondaryTableEntitySub b = new SecondaryTableEntitySub();
-        b.setB(4l);
-        b.setC(5l);
+        b.setB(4L);
+        b.setC(5L);
         em.persist(b);
 
         CriteriaBuilder<SecondaryTableEntitySub> criteria = cbf.create(em, SecondaryTableEntitySub.class, "d");
-        criteria.where("d.c").eq(5l);
+        criteria.where("d.c").eq(5L);
 
         assertFalse(criteria.getResultList().isEmpty());
     }

--- a/integration/hibernate-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/HibernateJpaProvider.java
+++ b/integration/hibernate-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/HibernateJpaProvider.java
@@ -506,12 +506,19 @@ public class HibernateJpaProvider implements JpaProvider {
         }
 
         // In this case, the property might represent a formula
-        if (columnNames.length == 1 && columnNames[0] == null) {
+        boolean isFormula = columnNames.length == 1 && columnNames[0] == null;
+        boolean isSubselect = tables.length == 1 && tables[0] == null;
+
+        if (isFormula || isSubselect) {
             Type propertyType = entityPersister.getPropertyType(attributeName);
             long length;
             int precision;
             int scale;
             try {
+                if (propertyType instanceof org.hibernate.type.EntityType) {
+                    propertyType = ((org.hibernate.type.EntityType) propertyType).getIdentifierOrUniqueKeyType(sfi);
+                }
+
                 Method m = Type.class.getMethod("dictatedSizes", Mapping.class);
                 Object size = ((Object[]) m.invoke(propertyType, sfi))[0];
                 length =    (long) size.getClass().getMethod("getLength").invoke(size);

